Adds Syndicate tablets, replaces the pinpointer used by Nuke Ops

zxaber opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
- Adds a new tablet subtype; a syndicate tablet. This tablet has access to the Syndicate software repository (and _only_ the Syndicate repository) in the software download app. It also comes with Fission360 pre-installed.

- Replaces the syndicate pinpointer for Nuke Ops with the new Syndicate tablet. Lone Ops, being a subtype of normal Nuke Ops, also get the tablet.

- Adds proper program overlays for the two radar apps (Fission360 and Lifeline), which shows the direction of the current target, a dot if they are on the same tile as the computer, or a miniature error bar if the signal was lost. This works even if the window is closed (assuming a target was selected at some point).

- Cleans up some radar code.

- A bunch of minor changes to a number of modular computer and modular app files that ultimately make Syndicate modular computers have appropriate themes on their main menu, hardware config, and file manager tguis. Also changes the word "NtOS" in the system bar to be dynamic, instead reading "Synux" on Syndicate devices.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Gives nuke ops access to the new Fission360 app, as well as the to-be-renamed Syndicate Borg management app and other minor apps in the syndicate repo.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Nuke Ops (and Lone Ops) now start with a Syndicate-themed tablet, rather than the pinpointer. The Syndicate tablet comes pre-installed with an app that replaces the function of disk pinpointers.
add: Lifeline and Fission360, the apps that track suit sensors and nuke disks respectively, now show the tracking status on the screen of the tablet/laptop object itself, without needing to keep the window open.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

![image](https://user-images.githubusercontent.com/37497534/83317967-b4230c80-a1e5-11ea-9f7f-2a3d6dde0e8a.png)
Syndicate tablets are visually distinct, and are resistant to emagging (as the one feature emagging grants is already unlocked).

zxaber committed:
Synux

zxaber committed:
cleanup

wesoda25 commented:
Do they have any other functions? I see it has access to the syndicate repo but I don’t really know what that entails. Also I think it’d be pretty cool if you could consolidate their uplinks into the tablet so that we can get rid of that, if it doesn’t have that function already.

zxaber commented:
Actually, moving uplinks into tablets is an obvious idea and I'm not sure why I didn't consider that.

Currently, the only other apps available are the cyborg monitor app (similar to the one available to the station but works on Syndicate borgs instead), an app that allows DDoS attacks against NTnet, and an app that allows for frying hard drives. I'm hoping that by exposing the tablets to use like this, other apps get developed.

BeeSting12 commented:
Does the pinpointer show up on the tablet screen while in your bag or do you have to have the window open to use it?

GunHog commented:
It does not, which is kind of a problem if you do not have the screen space to spare for the full program. I think that issue needs to be addressed before it can replace the pinpointer. The program window is quite big, and seems to disallow resizing.
![image](https://user-images.githubusercontent.com/4955510/83321286-27dc0e00-a214-11ea-86bd-0aebb240dfdd.png)

stylemistake commented:
Call it Syndix.

stylemistake commented on tgui/packages/tgui/layouts/NtosWindow.js:
```diff
@@ -17,6 +17,7 @@ export const NtosWindow = (props, context) => {
   } = props;
   const { act, data } = useBackend(context);
   const {
+    PC_device_theme,
```
Please override the theme here, in NtosWindow based on this variable, not in every single interface.

stylemistake replied:
You can do it like this (allow overriding with a "theme" prop, otherwise default to PC_device_theme):
```js
<Window
  theme={theme || PC_device_theme }
  resizable={resizable}>
```

zxaber replied:
Done.

zxaber replied:
I think this has been solved via explaining my intent (style, if not, let me know)

stylemistake commented on tgui/packages/tgui/layouts/NtosWindow.js:
```diff
@@ -41,7 +42,12 @@ export const NtosWindow = (props, context) => {
             </Box>
             <Box inline italic mr={2} opacity={0.33}>
-              NtOS
+              {PC_device_theme==="ntos"
+                ?"NtOS"
+                :""}
+              {PC_device_theme==="syndicate"
+                ?"Synux"
+                :""}
```
```suggestion
              {PC_device_theme === 'ntos' && 'NtOS'}
              {PC_device_theme === 'syndicate' && 'Syndix'}
```

zxaber replied:
done

stylemistake commented on code/modules/modular_computers/file_system/programs/ntdownloader.dm:
```diff
@@ -115,6 +123,7 @@
 
 	var/list/data = get_header_data()
 
+	data["theme"] = theme
```
Can you make it somewhat consistent, because you have an extra `theme` in addition to `PC_device_theme`? It's duplicate functionality, make it only `PC_device_theme`.

zxaber replied:
The reason I have these split is that I'm using `theme` when it's being set due to the program, and `PC_device_theme` when it's being set due to the type of computer. The downloader, file manager, and hardware config apps are all "built-in" apps (in that they cannot be deleted, transferred, or downloaded), and so ought to be following the theme set by the device. Contrast to individual apps like Fission360, which can be used on an NT device if it's been emagged.

The long and the short of it is that by using Fission360 on an NT tablet, you'd have `device_theme = "ntos"` but also `theme = "syndicate"`. The former would change the system tray text to say NtOS, while the latter tells the app how to look.

zxaber replied:
The big block of text here I meant to add to the first change request and got mixed up.

ntdownloader now uses `PC_device_theme`, as do the hardware config and file manager.

HomocidlePancakes commented:
I’d rather have a pinpointer than a tablet that will clutter up my screen.

zxaber commented:
I'll give Fission360 a dynamic screen with an arrow, which will function the same as the pinpointer does now. That way, as long as the program is the active one, you can close the window and just use the arrow like before.

HomocidlePancakes commented:
Sweet

kopoba commented:
can you see this new pinpointer if its in backpack? or it must be in pocket or in hand to see nuke disc?

zxaber commented:
The new pinpointer is a tablet app. As such, the gui can only be seen while it is held or in a pocket (or belt/ID slot). However, as per above requests, I'm currently working on setting up the tablet's sprite to have a visual arrow, working much the same way as the SM monitor app works. If the the disk tracking app (Fission360) is the active program, the tablet will work like the old pinpointers, even if the tgui window is closed.

ExcessiveUseOfCobblestone commented:
I'd just like to interject for moment. What you're refering to as Synux, is in fact, NTNet/Synux, or as I've recently taken to calling it, NTNet plus Synux. Synux is not an operating system unto itself, but rather another free component of a fully functioning NTNet system made useful by the NTNet corelibs, shell utilities and vital system components comprising a full OS as defined by POSIX.

Many computer users run a modified version of the NTNet system every day, without realizing it. Through a peculiar turn of events, the version of NTNet which is widely used today is often called Synux, and many of its users are not aware that it is basically the NTNet system, developed by the NTNet Project.

There really is a Synux, and these people are using it, but it is just a part of the system they use. Synux is the kernel: the program in the system that allocates the machine's resources to the other programs that you run. The kernel is an essential part of an operating system, but useless by itself; it can only function in the context of a complete operating system. Synux is normally used in combination with the NTNet operating system: the whole system is basically NTNet with Synux added, or NTNet/Synux. All the so-called Synux distributions are really distributions of NTNet/Synux.

stylemistake commented:
![sinus_infection_causes_tooth_pain_1_1](https://user-images.githubusercontent.com/1516236/83835902-ab0cd080-a6fa-11ea-96d7-352d14f30ebe.jpg)

zxaber committed:
I bet this merge conflicts

we hate tgui.bundle

zxaber commented:
Tablet screen updating is now done, more or less.
https://streamable.com/5ayerb

Arrow colors change at range 1/13/25 rather than 1/9/17, so as to sync up with the UI range (that is, the red arrow shows on the overlay at the same range that the off-screen arrow shows on the UI). Blue has been replaced with yellow in the tablet version to be more noticeable against the NTOS theme background (Used by Lifeline) and also to be easier to tell apart from the green close-range arrow.

While I was at it, I cleaned up some of the radar code.

zxaber committed:
extra bits

zxaber committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into SyndiTablets

zxaber committed:
Syndix

Twaticus reviewed (approved):
lookin good

zxaber committed:
small fixes

stylemistake reviewed (dismissed):
I assume this is all functional. Looks good.

zxaber committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into SyndiTablets

zxaber commented:
We are momentarily in-between merge conflicts.

Rohesie commented on code/modules/modular_computers/computers/item/computer_ui.dm:
```diff
@@ -39,13 +39,19 @@
 		assets.send(user)
-		ui = new(user, src, ui_key, "NtosMain", "NtOS Main menu", 400, 500, master_ui, state)
+		var/headername
+		if(device_theme == "ntos")
+			headername = "NtOS Main Menu"
+		if(device_theme == "syndicate")
+			headername = "Syndix Main Menu"
```
```suggestion
		switch(device_theme)
			if("ntos")
				headername = "NtOS Main Menu"
			if("syndicate")
				headername = "Syndix Main Menu"
```

zxaber replied:
Done.

Rohesie commented on code/modules/modular_computers/file_system/programs/ntdownloader.dm:
```diff
@@ -30,8 +37,8 @@
-	// Attempting to download antag only program, but without having emagged computer. No.
-	if(PRG.available_on_syndinet && !(computer.obj_flags & EMAGGED))
+	// Attempting to download antag only program, but without having emagged/syndicate computer. No.
+	if(PRG.available_on_syndinet && !((computer.obj_flags & EMAGGED) || (computer.device_theme = "syndicate")))
```
Could you make the syndicate computer property dissociated from the visual theme, and instead its own thing?

zxaber replied:
Rewrote this section to use a variable in the nt downloader rather than whether the system is emagged or uses syndicate style. Syndicate versions simply enable the emagged var by default.

Rohesie commented on code/modules/modular_computers/file_system/programs/radar.dm:
```diff
@@ -21,12 +21,29 @@
+	var/pointercolor = "green"
+
+/datum/computer_file/program/radar/run_program(mob/living/user)
+	if(..())
+		START_PROCESSING(SSfastprocess, src)
+		return TRUE
```
```suggestion
	. = ..()
	if(.)
		START_PROCESSING(SSfastprocess, src)
		return
```

zxaber replied:
Done.

Rohesie commented on code/modules/modular_computers/file_system/programs/radar.dm:
```diff
@@ -21,12 +21,29 @@
 	return ..()
 
+/datum/computer_file/program/radar/Destroy()
+	STOP_PROCESSING(SSfastprocess, src)
+	. = ..()
```
```suggestion
	STOP_PROCESSING(SSfastprocess, src)
	return ..()
```

zxaber replied:
Done.

Rohesie commented on code/modules/modular_computers/file_system/programs/radar.dm:
```diff
@@ -65,7 +82,37 @@
+		else if(target_turf.z < here_turf.z)
+			pointer="caret-down"
+
+	var/list/trackinfo = list(
+		locx = locx,
+		locy = locy,
+		userot = userot,
+		rot = rot,
+		arrowstyle = arrowstyle,
+		color = pointercolor,
+		pointer = pointer,
+		)
```
You might want to use string indexes, such as `"rot" = rot` and `"userot" = userot` else you'll have odd elements like `0 = 0` and whatnot, making this into an associative list which you can only access directly.

zxaber replied:
Done.

zxaber committed:
Summary (required)

zxaber commented:
Changed the modular computer emag function to be as follows:
If you emag the computer, it will call `run_emag()` on all programs stored on its drive. The default `run_emag()` proc will just return FALSE. Programs that have an emag function (currently, just the NT app downloader) will make note of the emag and return TRUE if emagging actually changes something. The computer will then tell the user if emagging did anything based via some (slightly ambiguous) fluff text.

Unlike most objects, the device can be emagged multiple times, in case they later download another program that has an emag function. The device will set its own emag flag the first time it gets emagged (for consistency), but the emag flag is not used by the programs and the device does not have any emag functions on its own. The NT Downloader app has been updated to reflect these changes.

--

As much as I liked my old overlay sprites, I just couldn't shake the feeling that they were overly busy for something you'd want to look at quickly. So I removed the little grid to make them easier to read at a glance.
![image](https://user-images.githubusercontent.com/37497534/85192262-b7566900-b276-11ea-96d9-307ec4fbeebb.png)

zxaber committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into SyndiTablets

zxaber committed:
this linter drives me up the wall

zxaber committed:
final touchups

Rohesie commented:
Tested locally and working?

zxaber commented:
I can't find anymore issues from my own tests.

Rohesie commented:
Merge condlict.
After it's solved, do you feel confident for it to go in, or would you like it to be tested a bit before?

zxaber committed:
Forgot this thing

zxaber committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into SyndiTablets

zxaber commented:
I ran through a bunch of testing again, and the only thing I could find was that I forgot to put in a check for whether the nukie tablet was powered on before attempting (and failing) emags.

I'd be open for a test merge but I've already fixed the same conflict like thrice now. Anyway, I think it's good.
